File: src/cli/commands/enable.ts
/**
 * Enable command implementation
 * 
 * Enables a host entry by hostname in ~/.hosts
 */

import { readHostsFile, writeHostsFile } from "../../core/file-io";
import type { HostsFile, Group } from "../../domain/types";
import { ExitCode } from "../exit-codes";

/**
 * Execute the enable command
 * 
 * @param hostname - Hostname to enable
 * @returns Exit code (0 = success, 1 = not found, 2 = I/O error)
 */
export async function enableCommand(hostname: string): Promise<number> {
  try {
    // Read ~/.hosts
    const hostsFile = await readHostsFile("~/.hosts");

    // Find and enable the entry
    let found = false;
    const updatedHostsFile = enableEntryByHostname(hostsFile, hostname, (wasFound) => {
      found = wasFound;
    });

    // If not found, return exit code 1
    if (!found) {
      console.error(`Error: Hostname '${hostname}' not found`);
      return ExitCode.VALIDATION;
    }

    // Write back to ~/.hosts
    await writeHostsFile("~/.hosts", updatedHostsFile);

    console.log(`✓ Enabled '${hostname}'`);
    return ExitCode.SUCCESS;

  } catch (err: any) {
    // Handle I/O errors
    console.error(`Error: ${err.message}`);
    return ExitCode.IO_ERROR;
  }
}

/**
 * Recursively enable an entry by hostname in a HostsFile
 * 
 * @param hostsFile - The hosts file structure
 * @param hostname - Hostname to enable
 * @param onFound - Callback to signal if entry was found
 * @returns Updated HostsFile with entry enabled
 */
function enableEntryByHostname(
  hostsFile: HostsFile,
  hostname: string,
  onFound: (found: boolean) => void
): HostsFile {
  let found = false;

  const updatedGroups = hostsFile.groups.map(group => 
    enableInGroup(group, hostname, (wasFound) => {
      if (wasFound) found = true;
    })
  );

  onFound(found);

  return {
    ...hostsFile,
    groups: updatedGroups,
  };
}

/**
 * Recursively enable an entry in a group and its nested groups
 * 
 * @param group - The group to process
 * @param hostname - Hostname to enable
 * @param onFound - Callback to signal if entry was found
 * @returns Updated Group with entry enabled
 */
function enableInGroup(
  group: Group,
  hostname: string,
  onFound: (found: boolean) => void
): Group {
  // Update entries in this group
  const updatedEntries = group.entries.map(entry => {
    if (entry.hostname === hostname) {
      onFound(true);
      return {
        ...entry,
        enabled: true,
      };
    }
    return entry;
  });

  // Recursively process nested groups
  const updatedNestedGroups = group.groups.map(nestedGroup =>
    enableInGroup(nestedGroup, hostname, onFound)
  );

  return {
    ...group,
    entries: updatedEntries,
    groups: updatedNestedGroups,
  };
}

File: go/internal/cmd/enable.go
package cmd

import (
	"fmt"

	"github.com/hungthai1401/hostie/go/internal/core/fileio"
	"github.com/hungthai1401/hostie/go/internal/domain"
	"github.com/spf13/cobra"
)

func newEnableCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "enable <hostname>",
		Short: "Enable a hosts entry",
		Long:  "Enable an entry in the hosts file by hostname",
		Args:  cobra.ExactArgs(1),
		RunE:  runEnable,
	}

	return cmd
}

func newDisableCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "disable <hostname>",
		Short: "Disable a hosts entry",
		Long:  "Disable an entry in the hosts file by hostname",
		Args:  cobra.ExactArgs(1),
		RunE:  runDisable,
	}

	return cmd
}

func runEnable(cmd *cobra.Command, args []string) error {
	return setEntryEnabled(args[0], true)
}

func runDisable(cmd *cobra.Command, args []string) error {
	return setEntryEnabled(args[0], false)
}

func setEntryEnabled(hostname string, enabled bool) error {
	// Read existing hosts file
	hostsFile, err := fileio.ReadHostsFile(hostsFilePath)
	if err != nil {
		return fmt.Errorf("failed to read hosts file: %w", err)
	}

	// Set entry enabled state
	found := setEntryEnabledByHostname(&hostsFile, hostname, enabled)
	if !found {
		return fmt.Errorf("entry not found: %s", hostname)
	}

	// Write back to file
	if err := fileio.WriteHostsFile(hostsFilePath, hostsFile); err != nil {
		return fmt.Errorf("failed to write hosts file: %w", err)
	}

	action := "enabled"
	if !enabled {
		action = "disabled"
	}
	fmt.Printf("✓ %s %s\n", hostname, action)
	return nil
}

// setEntryEnabledByHostname sets the enabled state of an entry by hostname
// Returns true if the entry was found, false otherwise
func setEntryEnabledByHostname(hf *domain.HostsFile, hostname string, enabled bool) bool {
	for i := range hf.Groups {
		if setGroupEntryEnabled(&hf.Groups[i], hostname, enabled) {
			return true
		}
	}
	return false
}

// setGroupEntryEnabled recursively sets the enabled state in a group and its subgroups
func setGroupEntryEnabled(g *domain.Group, hostname string, enabled bool) bool {
	// Check entries in this group
	for i := range g.Entries {
		if g.Entries[i].Hostname == hostname {
			g.Entries[i].Enabled = enabled
			return true
		}
	}

	// Check subgroups
	for i := range g.Groups {
		if setGroupEntryEnabled(&g.Groups[i], hostname, enabled) {
			return true
		}
	}

	return false
}
